#make files for and run RFMix
#python 04_RFmix.py --ref_bcf /home/ryan/Local_ancestry/admixture-simulation/ACB_simulation.ref.bcf.gz --query_vcf /home/ryan/Local_ancestry/admixture-simulation/ACB_simulation.query.vcf --pop admixture_simulation/ACB.txt --cleanup

import argparse
import os
import pandas as pd

#take in data to convert
parser = argparse.ArgumentParser()
parser.add_argument("--ref_bcf", type = str, action = "store", dest = "ref", required = True, help = "Reference panel generated by admixture-simulation.py. Ends with .ref.bcf.gz.")
parser.add_argument("--query_vcf", type = str, action = "store", dest = "query", required = True, help = "Query genotypes generated by admixture-simulation.py. Ends with .vcf")
parser.add_argument("--pop", type = str, action = "store", dest = "pop", required = True, help = "pop code file that was input into admixture-simulation.py")
parser.add_argument("--chr", type = str, action = "store", default = "22", dest = "chrom", required = False, help = "Chromosome of the input file. Default = 22.")
parser.add_argument("--tag", type = str, action = "store", dest = "tag", required = False, help = "additional tag for differentiating output say sample size or cohort ID")
parser.add_argument("--out", type = str, action = "store", default = "sim_RFMix", dest = "out", required = False, help = "output folder. Default = cwd/sim_RFMix")
parser.add_argument("--map", type = str, action = "store", dest = "map", required = True, help = "correct genetic map file SNP, BP, cM")
parser.add_argument("--cleanup", action = "store_true", default = False, required = False, help = "Include this flag to cleanup the directory after RFMix is done running to just leave results.")
args = parser.parse_args() #then pass these arguments to further things

arg_pop = args.pop
arg_pop_name = arg_pop.split("/")[-1].replace(".txt", "")
ref_bcf = args.ref
arg_tag = args.tag
out= args.out
chrom = args.chrom
map = args.map
os.system("mkdir -p " + out)
cwd = out + "/"

#split pop file and vcf into references 
pop = pd.read_csv(arg_pop, header = None, sep = "\t")
cohort_anc_pops = []
for ref_pop_name in ["NAT", "CEU", "YRI"]:
    ref_pop_IDs = pop.loc[pop[1] == ref_pop_name].drop(1, axis = 1)
    
    if len(ref_pop_IDs > 0): #if that pop is in the cohort
        cohort_anc_pops.append(ref_pop_name)
        ref_pop_IDs.to_csv(cwd + ref_pop_name + "_" + arg_pop_name + "_IDs.txt", index = False, header = False) #write reference id lists to file
    
        #pull reference genos and convert 
        os.system("bcftools view -S " + cwd + "/" + ref_pop_name + "_" + arg_pop_name + "_IDs.txt --force-samples -O z -o " + cwd + "/" + ref_pop_name + "_" + arg_pop_name + "_ref.vcf.gz " + ref_bcf)
        os.system("bcftools index --threads 40 -f --tbi " + cwd + "/" + ref_pop_name + "_" + arg_pop_name + "_ref.vcf.gz > " + cwd + "/" + ref_pop_name + "_" + arg_pop_name + "_ref.vcf.tbi")

#extract query and merge ref and query in the correct order (I now realize this is redundant but I already wrote it so whoops)... also this is useful for ELAI later
os.system("cp " + args.query + " " + cwd + "/" + arg_pop_name + ".vcf; bcftools view " + cwd + "/" + arg_pop_name + ".vcf -O z -o " + cwd + "/" + arg_pop_name + ".vcf.gz")
os.system("bcftools index --threads 40 -f --tbi " + cwd + "/" + arg_pop_name + ".vcf.gz > " + cwd + "/" + arg_pop_name + ".vcf.tbi")
cohort_anc_pops_str = ""
for cohort_anc_pop in cohort_anc_pops: #some of them only have two ancestries
    cohort_anc_pops_str = cohort_anc_pops_str + "" + cwd + "/" + cohort_anc_pop + "_" + arg_pop_name + "_ref.vcf.gz "
os.system("bcftools merge -Ov " + cohort_anc_pops_str + "" + cwd + "/" + arg_pop_name + ".vcf.gz -o " + cwd + "/" + arg_pop_name + "_merged_unfiltered.vcf")

#remove SNPs without genetic map coordinates
os.system("awk '{print $1}' " + map +  " > " + cwd + "/chr" + chrom + ".interpolated_genetic_map.SNPs")
os.system("vcftools --vcf " + cwd + "/" + arg_pop_name + "_merged_unfiltered.vcf --snps " + cwd + "/chr" + chrom + ".interpolated_genetic_map.SNPs --recode --out " + cwd + "/" + arg_pop_name + "_merged; mv " + cwd + "/" + arg_pop_name + "_merged.recode.vcf " + cwd + "/" + arg_pop_name + "_merged.vcf")
os.system("vcf-query -l " + cwd + "/" + arg_pop_name + "_merged.vcf > " + cwd + "/" + arg_pop_name + "_merged.vcf_ids.txt") #max missing has genotyping rate of 100%

#make classes file #arg_pop_name = "pop_codes_80_20"
vcf_IDs = pd.read_csv("" + cwd + "/" + arg_pop_name + "_merged.vcf_ids.txt", header = None, sep = "\t")
vcf_IDs.columns = [("IID")]
pop.columns = ["IID", "pop"]
vcf_pops = pd.merge(vcf_IDs, pop, on = "IID", how = "left")
vcf_pops["pop"] = vcf_pops["pop"].astype("category")
vcf_pops["pop"] = pd.factorize(vcf_pops["pop"])[0] + 1 #admixed is 0
with open("" + cwd + "/" + arg_pop_name + ".classes", "w") as f:
    for item in vcf_pops["pop"]:
        item_str = str(item) + " "
        f.write(item_str) #haplotypes so do it twice
        f.write(item_str)

#make haplotypes file
os.system("bcftools convert --hapsample --vcf-ids " + cwd + "/" + arg_pop_name + "_merged.vcf -o " + cwd + "/" + arg_pop_name + "_merged.haps")
os.system("zcat " + cwd + "/" + arg_pop_name + "_merged.haps.hap.gz | awk '{ $1=\"\"; $2=\"\"; $3=\"\"; $4=\"\"; $5=\"\"; print}' | sed 's/\s//g' | sed s/\\*//g > " + cwd + "/" + arg_pop_name + "_merged.haps")
os.system("sed '/##/d' " + cwd + "/" + arg_pop_name + "_merged.vcf | cut -f3 > " + cwd + "/" + arg_pop_name + ".snps")

#make cM map
vcf_SNPs = pd.read_csv("" + cwd + "/" + arg_pop_name + ".snps")
full_gen_map = pd.read_csv(map, header = None, sep = " ")
vcf_SNPs.columns = [("SNP")]
full_gen_map.columns = ["SNP", "BP", "CM"]
vcf_gen_map = pd.merge(vcf_SNPs, full_gen_map, on = "SNP", how = "left").drop("SNP", axis = 1).drop("BP", axis = 1)
vcf_gen_map.to_csv(cwd + "/" + arg_pop_name + ".pos", index = False, header = False, sep = "\t", na_rep = "NA")

#leggo (and time it while at it)
os.system("mkdir -p " + out + "/results_" + arg_tag + "/")
os.system("cd /home/rschubert1/scratch/software/RFMix_v1.5.4; (~/scratch/time --verbose python RunRFMix.py -e 2 -w 0.2 --num-threads 1 --forward-backward PopPhased " + out + "/" + arg_pop_name + "_merged.haps " + out + "/" + arg_pop_name + ".classes " + out + "/" + arg_pop_name + ".pos -o " + out + "/results_" + arg_tag + "/" + arg_pop_name + ".rfmix) 2> " + out + "/results_" + arg_tag + "/" + arg_pop_name + "_" + arg_tag + "_benchmarking.txt; cd " + out)

#rm extra stuff since results are in their own folder
if args.cleanup:
  os.system("rm " + cwd + "/*")

print("RFMix has completed running on " + arg_pop_name + ". Have a nice day :).")
